Stamus integration v1.0.0

regit opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues

Linked issue: https://github.com/demisto/content/issues/26285

## Description

This integration provides interaction with Stamus Security Platform. It provides a way to fetch detection from the solution and 
also getting information and context from the Stamus Security Platform.

## Screenshots

![Screenshot from 2023-05-03 20-07-15](https://user-images.githubusercontent.com/308616/236005952-adc5bb4d-0314-4777-b5b1-96ec66a9678f.png)

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [x] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

sonicold committed:
packs: add stamus pack to xsoar content

Implement a new set of commands to interact with
Stamus Security Platform.

regit committed:
Stamus: remove dummy test

regit committed:
Stamus: use StamusIntegration prefix

regit committed:
Stamus: fix some descriptions

regit committed:
Stamus: update result to return markdown

regit committed:
Stamus: rename commands

regit committed:
Stamus: update pack metadata

regit committed:
Stamus: homogeneize message

regit committed:
Stamus: really basic README

regit committed:
Stamus: fix some linters warning

regit committed:
Stamus: import playbook and mapper

regit committed:
Stamus: more import

Import Stamus' IncidentFields, IncidentFields and Layouts

regit committed:
Stamus: add playbook documentation

regit committed:
Stamus: more documentation

regit committed:
Stamus: add generated docs

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=26286) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=26286) before we can accept your contribution.<br/>**0** out of **3** committers have signed the CLA.<br/><br/>:x: sonicold<br/>:x: regit<br/>:x: pevma<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=26286) it.</sub>

GuyAfik commented:
> [![CLA assistant check](https://camo.githubusercontent.com/4c970efeed75dd01f6778bf7e33dec85e4171375cb69acb0927bfc9d437d78d9/68747470733a2f2f636c612d617373697374616e742e696f2f70756c6c2f62616467652f6e6f745f7369676e6564)](https://cla-assistant.io/demisto/content?pullRequest=26286) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=26286) before we can accept your contribution.**0** out of **2** committers have signed the CLA.❌ sonicold❌ regitYou have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=26286) it.

@sonicold @regit can you please sign the CLA so we can move forward with the review?

regit commented:
> > > [![CLA assistant check](https://camo.githubusercontent.com/4c970efeed75dd01f6778bf7e33dec85e4171375cb69acb0927bfc9d437d78d9/68747470733a2f2f636c612d617373697374616e742e696f2f70756c6c2f62616467652f6e6f745f7369676e6564)](https://cla-assistant.io/demisto/content?pullRequest=26286) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=26286) before we can accept your contribution.**0** out of **2** committers have signed the CLA.x sonicoldx regitYou have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=26286) it.
> > 
> > 
> > @sonicold @regit can you please sign the CLA so we can move forward with the review?
> 
> Hello, I did do it as a partner and both @sonicold and myself are working for Stamus Networks. Is it enough ?

OK, just saw that there is 2 different things.

GuyAfik commented:
> > > [![CLA assistant check](https://camo.githubusercontent.com/4c970efeed75dd01f6778bf7e33dec85e4171375cb69acb0927bfc9d437d78d9/68747470733a2f2f636c612d617373697374616e742e696f2f70756c6c2f62616467652f6e6f745f7369676e6564)](https://cla-assistant.io/demisto/content?pullRequest=26286) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=26286) before we can accept your contribution.**0** out of **2** committers have signed the CLA.x sonicoldx regitYou have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=26286) it.
> > 
> > 
> > @sonicold @regit can you please sign the CLA so we can move forward with the review?
> 
> Hello, I did do it as a partner and both @sonicold and myself are working for Stamus Networks. Is it enough ?

the CLA looks good now, thank you

GuyAfik commented on Packs/Stamus/Integrations/Stamus/Stamus.py:
```diff
@@ -0,0 +1,398 @@
+        arg_name='First fetch time',
+        required=True
+    )
+    if first_fetch_time is None:
+        first_fetch_time = datetime.now()
```
```suggestion
```

can it be `Nonez` looks like according to line 341 it cannot cause you have default value.

GuyAfik commented on Packs/Stamus/Integrations/Stamus/Stamus.py:
```diff
@@ -0,0 +1,398 @@
+
+            demisto.setLastRun(next_run)
+            demisto.incidents(incidents)
+            # return_results(incidents)
```
```suggestion
```

GuyAfik commented on Packs/Stamus/Integrations/Stamus/Stamus.yml:
```diff
@@ -0,0 +1,109 @@
+    execution: false
+    name: Stamus-Get-Host-Insight
+    outputs:
+    - contextPath: StamusIntegration.Output
```
as mentioned it not a good convention that all the commands would have the same context, each one should have different context output.

if you need any help for the names of the context outputs for each command, let me know

GuyAfik reviewed (changes requested):
@regit incredible work! a few notes:

1. each command should have a different context output, it seems like you implemented that each command will use the same context output which is not a good convention.
2. I see you approach to `dict` keys without the `get` method of the `dict` object, make sure to use `get` method to avoid un-necessary `KeyError` exceptions.
3. please see if you are able to fix the [build](https://app.circleci.com/jobs/github/demisto/content/698410) issues, most of them are validations.

please let me know if you need any help with anything, you can always reach me out in slack in DFIR.

regit committed:
Stamus: remove template comment

regit committed:
Stamus: use f-string

regit committed:
Stamus: simply test connection

regit committed:
Stamus: remove commented code

sonicold committed:
Stamus: docstring on escape function

sonicold committed:
Stamus: fix to have one context for each command

sonicold committed:
Stamus: refactoring

sonicold committed:
Stamus: remove useless code

sonicold commented on Packs/Stamus/Integrations/Stamus/Stamus.py:
```diff
@@ -0,0 +1,398 @@
+''' HELPER FUNCTIONS '''
+
+
+def escape(string):
```
It is used to escape other elasticsearch reserved characters

sonicold committed:
Stamus: get default value in results if no results

regit committed:
Stamus: run demisto format on yaml files

regit committed:
Stamus: update README

regit committed:
Stamus: fix missing description and format

regit committed:
Stamus: fix following validation

regit committed:
Stamus: add missing playbook image

regit committed:
Stamus: remove marketplacev2 support

regit committed:
Stamus: fix info extraction in get DoC command

efelmandar reviewed (changes requested):
Hi @regit Thank you for your contribution. I have reviewed the content in your PR and I appreciate the effort you have put into it. However, I noticed a few areas that require attention. Please see my detailed comments below:

1. Incident Field Configuration:
I noticed that the incident field `Stamus Threat` is currently configured to apply to "All Incident types", whereas all the other incident fields are configured specifically for the `Stamus Networks` incident type. Could you please clarify if this change was intentional? If it was, I kindly request you provide some insights into the reasoning behind this modification in the PR description or a comment. However, if this was unintentional, please update the configuration to be specific to the `Stamus Networks` incident type.

![IncidentField_StamusThreat](https://github.com/demisto/content/assets/102903097/67c3f4e8-0324-4f64-ab8a-3018242fcdaa)

2. Missing Playbook Description:
The playbook "Stamus Networks - Get Extra Data" appears to be missing a description. Including a descriptive explanation of the playbook's purpose and functionality is crucial for users to understand and effectively utilize the content. please add a meaningful description to the playbook.

3. Non-Existing Command Usage:
I noticed that the playbook is utilizing non-existing commands. While I understand that these commands may be implemented in the integration source code, it is important to declare them properly in the YAML file with inputs, outputs, and other relevant details. I kindly request you to either add the missing command declaration, including the required specifications, or alternatively, modify the playbook to utilize existing commands that are already declared.

I appreciate your attention to these matters. Please make the necessary updates to address the mentioned issues.
If you have any questions or need further clarification on any of the points mentioned above, please don't hesitate to reach out here or on Slack.

GuyAfik commented on Packs/Stamus/Integrations/Stamus/Stamus.yml:
```diff
@@ -0,0 +1,109 @@
+    execution: false
+    name: Stamus-Get-DoC-Events
+    outputs:
+    - contextPath: StamusIntegration.RelatedEvents
+      description: '[Get events for a Declaration of Compromise.]'
+      type: String
```
please add full context paths for each api response field if possible.

you can take a look at other integrations to see examples.

regit replied:
I'm struggling with this.

For example, the stamus-get-host-insight command is giving me a result similar to this when displayed:
![image](https://github.com/demisto/content/assets/308616/93603296-3354-45ee-83dc-b7a06e57e735)

This is an history of a host with all first time we have seen a service, an tls/http agent seen on the IP, an hostname, ...

I got some question:
- Should I put the timestamp in the contextPath ?
- Is the data structure correct for XSOAR ? I can not really set a contextPath to an event because it is a combination of `type` and `value`. Like `type:hostname` will potentially give me an interesting evidence but it is completely different of an information like `type:service` which tell me a service has been seen

GuyAfik replied:
* yes, you can definitely put timestamp in the contextPath in the yml, the type should be `Date`
*  this data structure looks good to me, you can explain very briefly in the description of each context path what are the possible informations as long as they are on the same contextPath. if your api response would look something like this for example:

```
[
    {
        "timestamp": "some timestamp"
        "ip": "some ip",
        "type": "some type"
        "value": some value
    }
]
```

then your context path would look something like:

```
    outputs:
    - contextPath: Stamus.HostInsight.timestamp
      description: "some description what it means.."
      type: Date
   - contextPath: Stamus.HostInsight.ip
      description: "some description what it means.."
      type: String
   - contextPath: Stamus.HostInsight.type
      description: "some description what it means.."
      type: String
   - contextPath: Stamus.HostInsight.value
      description: "some description what it means.."
      type: String
```

GuyAfik reviewed (changes requested):
@regit nice work! gave some more comments

please let me know if you need anything.

You can also remove the `Pipfile` and `Pipfile.lock` as those files are not needed.

GuyAfik reviewed (changes requested):
@regit nice work! gave some more comments

please let me know if you need anything.

You can also remove the `Pipfile` and `Pipfile.lock` as those files are not needed.

regit committed:
Stamus: use getter function

Co-authored-by: Guy Afik <53861351+GuyAfik@users.noreply.github.com>

regit committed:
Stamus: command should be lower case

regit committed:
Stamus: avoid double read of event

regit committed:
Stamus: little optimization

regit committed:
Stamus: add docstring

regit committed:
Stamus: remove not needed files

regit committed:
Stamus: add basic integration doc

GuyAfik commented:
@regit let me know when i can re-review? for any help feel free to reach me out

regit committed:
Stamus: fix JSON format

regit committed:
Stamus: fix some name in playbook

regit committed:
Stamus: rename IOC params

sonicold committed:
Stamus: fix demisto sdk validation

validation: demisto-sdk validate -i Packs/Stamus/IncidentTypes/incidenttype-Stamus_Networks.json
validation: fix: demisto-sdk format -i /home/snuser/cortex/Packs/Stamus/IncidentTypes/incidenttype-Stamus_Networks.json

sonicold committed:
Stamus: fix params name gotten from conf

sonicold committed:
Stamus: fix stamus extra data playbook following new naming

regit committed:
Stamus: fix playbook ID

GuyAfik commented:
@regit nice work! can you please refer to the old comments i gave you please?

* https://github.com/demisto/content/pull/26286#discussion_r1190718829
* https://github.com/demisto/content/pull/26286#discussion_r1190717114
* https://github.com/demisto/content/pull/26286#pullrequestreview-1420860907

in addition, can you please merge from demisto master to re-trigger the build? for any help you can speak to me.

pevma committed:
Stamus: Add descriptions

sonicold committed:
Stamus: move constants with other constants

sonicold committed:
Stamus: use getter function + fix linter

sonicold committed:
Stamus: use fstring

sonicold committed:
Stamus: unit tests impl

regit commented:
> @regit let me know when i can re-review? for any help feel free to reach me out

Updated the MR by addressing comments, adding unittests and rebasing on master.

GuyAfik commented on Packs/Stamus/Integrations/Stamus/Stamus_test.py:
```diff
@@ -0,0 +1,205 @@
+"""Base Integration for Cortex XSOAR - Unit Tests file
+
+Pytest Unit Tests: all funcion names must start with "test_"
+
+More details: https://xsoar.pan.dev/docs/integrations/unit-testing
+
+MAKE SURE YOU REVIEW/REPLACE ALL THE COMMENTS MARKED AS "TODO"
+
+You must add at least a Unit Test function for every XSOAR command
+you are implementing with your integration
+"""
```
```suggestion
```

GuyAfik reviewed (commented):
looks good, the only thing left is to fix the build issues that we have:

[here](https://circleci.com/gh/demisto/content/712599) and [here](https://app.circleci.com/pipelines/github/demisto/content/286432/workflows/d9fccf7c-3d5b-4c89-80e9-6778cca855fc/jobs/712603)

regit committed:
Stamus: remove template doc

Co-authored-by: Guy Afik <53861351+GuyAfik@users.noreply.github.com>

regit committed:
Stamus: add test data to secret-ignore

regit committed:
Stamus: fix event type

regit committed:
Stamus: update demisto docker image

regit committed:
Stamus: fix reference in classifier

regit committed:
Stamus: update README.md

Runned demisto-sdk generate-docs -i Packs/Stamus/Integrations/Stamus/Stamus.yml
and fixed the version string.

regit committed:
Stamus: fix linter warning

regit committed:
Stamus: fix linter error

efelmandar commented:
@regit Thank you again for your contribution, I see you fixed and pushed many changes so good job, but I still see that the playbook is using commands which don't seem to be implemented in the integration yet, please make sure you only use existing and implemented integration commands. or if it's about to be added in the future consider removing the playbook from this PR and add it back when the commands are added

regit commented:
> @regit Thank you again for your contribution, I see you fixed and pushed many changes so good job, but I still see that the playbook is using commands which don't seem to be implemented in the integration yet, please make sure you only use existing and implemented integration commands. or if it's about to be added in the future consider removing the playbook from this PR and adding it back when these commands are added
> 
> Again, please feel free to reach out here or on Slack if you have any other questions or issues, I'd be happy to try and help

Ooops, once more :(

We're going to handle that tomorrow.

regit commented:
> > @regit Thank you again for your contribution, I see you fixed and pushed many changes so good job, but I still see that the playbook is using commands which don't seem to be implemented in the integration yet, please make sure you only use existing and implemented integration commands. or if it's about to be added in the future consider removing the playbook from this PR and adding it back when these commands are added
> > Again, please feel free to reach out here or on Slack if you have any other questions or issues, I'd be happy to try and help
> 
> Ooops, once more :(
> 
> We're going to handle that tomorrow.

@efelmandar We have just checked this morning and it looks good to us. Can you point us the specific command or problem ? (a demisto-sdk command will help ;)

efelmandar commented:
@regit Thank you for taking the time to double check that, now the playbook uses the right commands and everything looks fine.
Thank you again for your efforts! it is most appreciated!

@GuyAfik approved on my end FYI

GuyAfik commented:
@regit looks good, lets do a demo, only left to fix linter issues (see the build failures)

regit committed:
Stamus: fix a playbook param

regit committed:
Stamus: fix key value

regit committed:
Stamus: set default value for incident

regit committed:
Stamus: update from version

regit committed:
Stamus: set default incoming mapper

regit committed:
Stamus: improve layout

regit committed:
Stamus: some more layout work

regit committed:
Stamus: use already defined fields

regit committed:
Stamus: add host first seen

regit committed:
Stamus: don't extract indicators

Most fields are information and we should not extract indicators.

regit committed:
Stamus: display threat info in layout

regit committed:
Stamus: don't try to display removed field

regit committed:
Stamus: add host insight info in incident layout

This is the first pass, we need to add the role.

regit committed:
Stamus: display major host insight information

regit committed:
Stamus: retry policy and fix a field

Host insight info can be available after some time if ever the
host has not be seen before.

GuyAfik commented:
@regit please pull from master to re-trigger the build, and fix any issue if there is after that.
if you need any help let me know

regit committed:
Stamus: update to fix validation

regit committed:
Stamus: add description to mapper

regit committed:
Stamus: add integration as 'start time' user

GuyAfik commented:
> @GuyAfik I'm struggling with this one, I don't know what I am missing https://app.circleci.com/pipelines/github/demisto/content/298224/workflows/27794974-779e-42a7-ae37-c266f05a0377/jobs/717860?invite=true#step-119-11335_250

@regit thanks for all the massive effort that you did on this PR, Ill merge it and we will handle the issues internally.
Great work!

regit commented:
> > @GuyAfik I'm struggling with this one, I don't know what I am missing https://app.circleci.com/pipelines/github/demisto/content/298224/workflows/27794974-779e-42a7-ae37-c266f05a0377/jobs/717860?invite=true#step-119-11335_250
> 
> @regit thanks for all the massive effort that you did on this PR, Ill merge it and we will handle the issues internally.
> Great work!

Thanks a lot for all the help from the Palo Alto team on this.
